subtests/docker_cli/login/login.py

```diff
@@ -0,0 +1,333 @@
+        return dc.execute()
+
+    def _check_credentials(self):
+        """
+        Examine saved credentials in docker auth file.
+        If this is a logout test, make sure the server key is absent.
+        Otherwise, make sure there's an auth entry for this server,
+        and its value contains the expected username + password.
+        """
+        auth_file = self.config['docker_auth_file']
+        fp = open(auth_file, 'rb')
+        creds = json.load(fp)
+        server = self.sub_stuff['server']
+
+        # Use test name to determine whether cred should be present or not
+        if 'logout' in self.__class__.__name__:
+            self.failif(server in creds['auths'],
+                        ('"%s" should not be in %s after docker logout'
+                         % (server, auth_file)))
+        else:
+            # Stored credentials
+            cred_hash = creds['auths'][server]['auth']
+            # ...simply a base64-encoded('username:password'), see:
+            #  https://coreos.com/os/docs/latest/registry-authentication.html
+            expected = b64encode('{user}:{passwd}'.format(**self.sub_stuff))
+            self.failif_ne(cred_hash, expected,
+                           'saved credentials in %s' % auth_file)
+
+    def docker_push(self, img_name, tag_name):
+        """
+        Run docker tag, then docker push against our local registry.
+        Tag should always succeed but make no assumptions about push.
+        """
+        push_image = self.config['push_image']
+        dc = DockerCmd(self, 'pull', [push_image])
+        mustpass(dc.execute())
+        self.sub_stuff['my_images'] = [push_image]
+        # FIXME: add push_image to self.containers_needing_cleanup
+
+        self.sub_stuff['img_name'] = img_name
+        self.sub_stuff['tag_name'] = tag_name
+        pushed_name = ('{server}/{img_name}:{tag_name}'
+                       .format(**self.sub_stuff))
+        dc = DockerCmd(self, 'tag', [push_image, pushed_name])
+        mustpass(dc.execute())
+        self.sub_stuff['my_images'] += [pushed_name]
+
+        return DockerCmd(self, 'push', [pushed_name]).execute()
+
+    def cleanup(self):
+        DockerContainers(self).clean_all(self.sub_stuff['my_containers'])
+        if 'my_images' in self.sub_stuff:
+            DockerImages(self).clean_all(self.sub_stuff['my_images'])
+        self._restore_docker_auth_file()
+
+
+class login_ok(login_base):
+
+    """
+    1. Run 'docker login' with correct password
+    2. Verify zero (success) exit status, and 'Login Succeeded' in stdout
+    """
+
+    def run_once(self):
+        super(login_ok, self).run_once()
+        self.sub_stuff["cmdresult"] = self.docker_login()
+
+    def postprocess(self):
+        super(login_ok, self).postprocess()
+        cmdresult = self.sub_stuff['cmdresult']
+        OutputGood(cmdresult)
+        mustpass(cmdresult)
+        self.failif_not_in("Login Succeeded", cmdresult.stdout,
```

Comment:
Oh, okay, as long as you thought about it.  No, it doesn't bother me too much now, but it will when it breaks :grinning:  Since you're already checking for a clean exit, it might even be okay to "increase by trimming, the intersection angles of the three adjoining surfaces", by writing ``self.failif_not_in("ucce", ...``.  IMHO, that four-letter combination is unlikely to appear in any failure message :grinning: